intellij/src/saros/intellij/ui/actions/ConnectServerAction.java

```diff
@@ -51,37 +59,123 @@ public void execute() {
 
     if (account == null) return;
 
-    connectAccount(account);
+    processConnectionRequest(account);
+  }
+
+  /**
+   * Connects to the given XMPP account after checking the validity of the connection request.
+   *
+   * <p>If the user is already connected to an account, the request is dropped if requested account
+   * matches the connected account.
+   *
+   * <p>Otherwise, if the user is already connected to an account and in an active session, the user
+   * is questioned whether they really want to change accounts. If this is confirmed by the user,
+   * the session is left in the process of re-connecting with the new account.
+   *
+   * @param account the requested account to connect to
+   */
+  private void processConnectionRequest(final XMPPAccount account) {
+    JID currentJID = connectionHandler.getLocalJID();
+    JID newJID = new JID(account.getUsername(), account.getDomain());
+
+    if (currentJID == null) {
+      connectToAccount(account, false);
+
+      return;
+    }
+
+    if (newJID.strictlyEquals(currentJID)) {
+      LOG.debug(
+          "Ignoring connection action as account is already connected. current JID: "
+              + currentJID
+              + ", new JID: "
+              + newJID);
+
+      return;
+    }
+
+    ISarosSession sarosSession = sarosSessionManager.getSession();
+
+    if (sarosSession == null) {
+      connectToAccount(account, true);
+
+      return;
+    }
+
+    String message = Messages.ConnectServerAction_leave_session_confirmation_message;
+
+    if (sarosSession.isHost()) {
+      message += Messages.ConnectServerAction_leave_session_confirmation_host_addendum_message;
+    }
+
+    boolean reallyReconnect;
+
+    try {
+      reallyReconnect =
+          SafeDialogUtils.showYesNoDialog(
+              project, message, Messages.ConnectServerAction_leave_session_confirmation_title);
+
+    } catch (IllegalAWTContextException e) {
+      LOG.error("Failed to show question on whether to really switch accounts to user", e);
+
+      NotificationPanel.showError(
+          MessageFormat.format(
+              Messages.ConnectServerAction_leave_session_confirmation_host_addendum_error_message,
+              e.getMessage()),
+          Messages.ConnectServerAction_leave_session_confirmation_host_addendum_error_title);
+
+      return;
+    }
+
+    if (reallyReconnect) {
```

Comment:
The magic is done somewhere else, however calling SessionManage#stop(in a background thread should be more understandable, do not we do this in Saros/E or do we rely on the listener in the SarosSessionManager too ?)

Reply:
I ended the session manually in my first draft but from I talk I got the impression that this was not needed/wanted. But adding it back in is no problem either.

And in Saros/E, this is also just handled through the disconnect. See [`XMPPConnectionSupport.connect(XMPPAccount account, boolean setAsDefault, boolean failSilently)`](https://github.com/saros-project/saros/blob/master/eclipse/src/saros/ui/util/XMPPConnectionSupport.java#L95).